users/service.go

```diff
@@ -109,4 +153,51 @@ func (svc usersService) UserInfo(ctx context.Context, token string) (User, error
 		Password: "",
 		Metadata: dbUser.Metadata,
 	}, nil
+
+}
+
+func (svc usersService) GenerateResetToken(ctx context.Context, email, host string) error {
+	user, err := svc.users.RetrieveByID(ctx, email)
+	if err != nil || user.Email == "" {
+		return ErrUserNotFound
+	}
+
+	tok, err := svc.token.Generate(email, 0)
+	if err != nil {
+		return ErrGeneratingResetToken
+	}
+	return svc.SendToken(ctx, host, email, tok)
+}
+
+func (svc usersService) UpdatePassword(ctx context.Context, token, password string) error {
+
+	email, err := svc.token.Verify(token)
+	if err != nil {
+		return err
+	}
+
+	u, err := svc.users.RetrieveByID(ctx, email)
+	if err != nil || u.Email == "" {
+		return ErrUserNotFound
+	}
+
+	password, err = svc.hasher.Hash(password)
+	if err != nil {
+		return err
+	}
+	return svc.users.UpdatePassword(ctx, email, password)
+}
+
+// SendToken sends password recovery link to user
+func (svc usersService) SendToken(_ context.Context, host, email, token string) error {
+	url := svc.email.ResetURL
+	msg := []byte(fmt.Sprintf("To: %s\r\n"+
+		"Subject: Reset Password!\r\n"+
+		"\r\n"+
+		"You have initiated password reset.\r\n"+
+		"Follow the link below to reset password.\r\n"+
+		"%s%s?token=%s", email, host, url, token))
```

Comment:
I think that this message should be constructed in the `email` package as part of the emailer.